File: Daily Questions/0567-permutation-in-string/0567-permutation-in-string.py
class Solution:
    def checkInclusion(self, pattern: str, text: str) -> bool:
        """
        Check if a permutation of the pattern is a substring of the text.

        Parameters:
        pattern (str): The pattern string for which we need to check permutations.
        text (str): The text string in which we check for permutations of the pattern.

        Returns:
        bool: True if a permutation of pattern exists as a substring in text, False otherwise.
        """
        pattern_length = len(pattern)
        text_length = len(text)

        # If the pattern is longer than the text, no permutation can be a substring
        if pattern_length > text_length:
            return False

        # Frequency counts of characters in pattern and the first window in text
        pattern_count = [0] * 26  # For storing frequency of each character in pattern
        window_count = [0] * 26   # For storing frequency of characters in current window of text

        # Fill the frequency count for the pattern and the first window of the text
        for i in range(pattern_length):
            pattern_count[ord(pattern[i]) - ord('a')] += 1
            window_count[ord(text[i]) - ord('a')] += 1

        # If the frequency of characters matches, the current window is a permutation
        if pattern_count == window_count:
            return True

        # Slide the window over the text, one character at a time
        for i in range(pattern_length, text_length):
            # Add the new character (right side of the window) to the window count
            window_count[ord(text[i]) - ord('a')] += 1
            
            # Remove the character that's left out of the window (left side of the window)
            window_count[ord(text[i - pattern_length]) - ord('a')] -= 1

            # Compare the pattern's character frequency with the current window's frequency
            if pattern_count == window_count:
                return True

        return False

File: Daily Questions/0567-permutation-in-string/0567-permutation-in-string.c
#include <stdbool.h>
#include <string.h>

/**
 * @brief Check if a permutation of s1 exists as a substring of s2.
 *
 * @param s1 (char*): The pattern string whose permutation we are checking.
 * @param s2 (char*): The text string where we are searching for the permutation of s1.
 *
 * @return bool: Returns true if a permutation of s1 exists in s2, otherwise false.
 */
bool checkInclusion(char *s1, char *s2) {
    int s1_length = strlen(s1);
    int s2_length = strlen(s2);

    // If s1 is longer than s2, no permutation can be a substring
    if (s1_length > s2_length) {
        return false;
    }

    // Arrays to store frequency counts of characters in s1 and current window in s2
    int s1_count[26] = {0};
    int s2_count[26] = {0};

    // Initialize frequency count arrays for s1 and the first window in s2
    for (int i = 0; i < s1_length; i++) {
        s1_count[s1[i] - 'a']++;
        s2_count[s2[i] - 'a']++;
    }

    // Compare s1's character frequency with the first window in s2
    if (memcmp(s1_count, s2_count, sizeof(s1_count)) == 0) {
        return true;
    }

    // Slide the window over s2, one character at a time
    for (int i = s1_length; i < s2_length; i++) {
        // Add the new character (right side of the window) to s2_count
        s2_count[s2[i] - 'a']++;

        // Remove the character that's no longer in the window (left side)
        s2_count[s2[i - s1_length] - 'a']--;

        // Compare the frequency arrays of s1 and current window of s2
        if (memcmp(s1_count, s2_count, sizeof(s1_count)) == 0) {
            return true;
        }
    }

    return false;
}
